sdk/pinocchio/src/cpi.rs

```diff
@@ -13,137 +13,269 @@ use crate::{
+/// consuming CUs in the process. The `invoke_signed_unchecked` is an alternative
+/// to this function that have lower CU consumption since it does not perform
+/// any validation. This should only be used when the caller is sure that the borrow
+/// checker rules are followed.
+///
+/// Note that the maximum number of accounts that can be passed to a cross-program
+/// invocation is defined by the `MAX_CPI_ACCOUNTS` constant. Even if the slice
+/// of `AccountInfo`s has more accounts, only the number of accounts required by
+/// the `instruction` will be used.
+///
 /// # Important
 ///
 /// The accounts on the `account_infos` slice must be in the same order as the
-/// `accounts` field of the `instruction`.
-#[inline]
+/// `accounts` field of the `instruction`. When the instruction has duplicated
+/// accounts, it is necessary to pass a duplicated reference to the same account
+/// to maintain the 1:1 relationship between `account_infos` and `accounts`.
 pub fn slice_invoke_signed(
     instruction: &Instruction,
     account_infos: &[&AccountInfo],
     signers_seeds: &[Signer],
 ) -> ProgramResult {
-    if instruction.accounts.len() < account_infos.len() {
+    // Check that the potential number of accounts provided is not less than
+    // the number of accounts expected by the instruction;
+    //
+    // The check for the slice of `AccountInfo`s not being less than the
+    // number of accounts expected by the instruction is done in
+    // `invoke_signed_with_bounds`.
+    if MAX_CPI_ACCOUNTS < instruction.accounts.len() {
         return Err(ProgramError::NotEnoughAccountKeys);
     }
 
-    if account_infos.len() > MAX_CPI_ACCOUNTS {
-        return Err(ProgramError::InvalidArgument);
-    }
-
-    const UNINIT: MaybeUninit<Account> = MaybeUninit::<Account>::uninit();
-    let mut accounts = [UNINIT; MAX_CPI_ACCOUNTS];
-    let mut len = 0;
-
-    for (account_info, account_meta) in account_infos.iter().zip(instruction.accounts.iter()) {
-        if account_info.key() != account_meta.pubkey {
-            return Err(ProgramError::InvalidArgument);
-        }
-
-        let state = if account_meta.is_writable {
-            BorrowState::Borrowed
-        } else {
-            BorrowState::MutablyBorrowed
-        };
-
-        if account_info.is_borrowed(state) {
-            return Err(ProgramError::AccountBorrowFailed);
-        }
+    _invoke_signed_with_bounds::<MAX_CPI_ACCOUNTS>(instruction, account_infos, signers_seeds)
+}
 
-        // SAFETY: The number of accounts has been validated to be less than
-        // `MAX_CPI_ACCOUNTS`.
-        unsafe {
-            accounts
-                .get_unchecked_mut(len)
-                .write(Account::from(*account_info));
-        }
+/// Internal function to invoke a cross-program instruction with signatures
+/// from a slice of `AccountInfo`s performing borrow checking.
+#[inline(always)]
+fn _invoke_signed_with_bounds<const MAX_ACCOUNTS: usize>(
```

Comment:
In Rust, symbol names that start with an underscore are excluded from the reachability analysis.
So you should not use it for things that are actually used.

I would actually argue this function is `unsafe`, as there is away to call it that would cause undefined behavior.  And this is exactly what `unsafe` is supposed to mark.
So, I think, it should be just an unsafe function, and the safety comment would explain the relationship between `MAX_ACCOUNTS` and both `instruction.accounts.len()` and `accounts_infos.len()`.

Reply:
Cool, fixed that. Also, improve the comments on the invoke functions. This is the commit [56b0b54](https://github.com/anza-xyz/pinocchio/pull/154/commits/56b0b54f19379c00758bbf53d26f57aa3c3f42f8).